File: api/utils/firebase_auth.py
"""Firebase authentication utilities for verifying user tokens."""
import logging
from typing import Optional

import firebase_admin
from fastapi import Depends, Header, HTTPException, status
from firebase_admin import auth, credentials

logger = logging.getLogger(__name__)

# Initialize Firebase Admin SDK (call once at startup)
_app: Optional[firebase_admin.App] = None


def init_firebase_app():
    """Initialize Firebase Admin SDK with default credentials."""
    global _app
    if _app is None:
        try:
            # Uses GOOGLE_APPLICATION_CREDENTIALS environment variable
            # or default credentials from the environment
            _app = firebase_admin.initialize_app()
            logger.info("Firebase Admin SDK initialized successfully")
        except Exception as e:
            logger.warning(f"Firebase Admin SDK initialization failed: {e}")
            logger.warning("User authentication will not be available")


async def get_firebase_uid(authorization: Optional[str] = Header(None)) -> Optional[str]:
    """
    Extract and verify Firebase ID token from Authorization header.

    Returns:
        Firebase UID if token is valid, None if no token provided

    Raises:
        HTTPException 401 if token is invalid
    """
    if not authorization:
        return None

    if not authorization.startswith("Bearer "):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid authorization header format. Expected 'Bearer <token>'",
        )

    token = authorization[7:]  # Remove "Bearer " prefix

    try:
        decoded_token = auth.verify_id_token(token)
        return decoded_token["uid"]
    except auth.InvalidIdTokenError:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid or expired Firebase token",
        )
    except Exception as e:
        logger.error(f"Firebase token verification failed: {e}")
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Token verification failed",
        )


# Dependency for endpoints that require authentication
FirebaseUidDependency = Depends(get_firebase_uid)

File: backend/src/utils/firebase-auth.ts
import * as admin from 'firebase-admin';
import { FastifyRequest, FastifyReply } from 'fastify';
import { config } from '../config';

let app: admin.app.App | null = null;
let isInitialized = false;

export function isFirebaseInitialized(): boolean {
  return isInitialized;
}

export function initFirebase(): void {
  if (app) return;

  try {
    // Read Firebase credentials from environment variable (JSON string)
    if (config.firebaseServiceAccount) {
      const serviceAccount = JSON.parse(config.firebaseServiceAccount);

      // Normalize private_key format - ensure proper line breaks
      if (serviceAccount.private_key && typeof serviceAccount.private_key === 'string') {
        let key = serviceAccount.private_key;

        // If the key has literal \n strings, replace them with actual newlines
        if (key.includes('\\n')) {
          key = key.replace(/\\n/g, '\n');
        }

        // Ensure the key starts and ends correctly
        key = key.trim();
        if (!key.startsWith('-----BEGIN PRIVATE KEY-----')) {
          throw new Error('Invalid private key format: missing BEGIN marker');
        }
        if (!key.endsWith('-----END PRIVATE KEY-----')) {
          throw new Error('Invalid private key format: missing END marker');
        }

        serviceAccount.private_key = key;
      }

      app = admin.initializeApp({
        credential: admin.credential.cert(serviceAccount),
        projectId: serviceAccount.project_id,
      });
    } else {
      // Auto-detect Render Secret File if available
      const renderSecretPath = '/etc/secrets/firebase-key.json';
      // We need to use require('fs') dynamically or import it at the top. 
      // Since we are in a try block, let's use a simple check if we can.
      // Better to rely on the standard GOOGLE_APPLICATION_CREDENTIALS if set, 
      // but if not, and the file exists, set it.

      if (!process.env.GOOGLE_APPLICATION_CREDENTIALS) {
        try {
          // eslint-disable-next-line @typescript-eslint/no-var-requires
          const fs = require('fs');
          if (fs.existsSync(renderSecretPath)) {
            console.log(`✅ Detected Render Secret File at ${renderSecretPath}`);
            process.env.GOOGLE_APPLICATION_CREDENTIALS = renderSecretPath;
          }
        } catch (e) {
          // Ignore fs errors
        }
      }

      // Fallback to GOOGLE_APPLICATION_CREDENTIALS file path if available
      app = admin.initializeApp({
        credential: admin.credential.applicationDefault(),
      });
    }
    console.log('✅ Firebase Admin SDK initialized successfully');
    isInitialized = true;
  } catch (error) {
    console.warn('⚠️  Firebase Admin SDK initialization failed:', error);
    console.warn('⚠️  User authentication will not be available');
    isInitialized = false;
  }
}

export async function verifyFirebaseToken(
  request: FastifyRequest,
  reply: FastifyReply
): Promise<string | null> {
  const authorization = request.headers.authorization;

  if (!authorization) {
    return null;
  }

  if (!authorization.startsWith('Bearer ')) {
    reply.code(401).send({
      error: 'Unauthorized',
      message: "Invalid authorization header format. Expected 'Bearer <token>'",
    });
    throw new Error('Invalid authorization header');
  }

  const token = authorization.substring(7); // Remove "Bearer " prefix

  try {
    const decodedToken = await admin.auth().verifyIdToken(token);
    return decodedToken.uid;
  } catch (error: any) {
    if (error.code === 'auth/id-token-expired') {
      reply.code(401).send({
        error: 'Unauthorized',
        message: 'Firebase token expired',
      });
    } else if (error.code === 'auth/argument-error') {
      reply.code(401).send({
        error: 'Unauthorized',
        message: 'Invalid Firebase token',
      });
    } else {
      console.error('Firebase token verification failed:', error);
      reply.code(401).send({
        error: 'Unauthorized',
        message: `Token verification failed: ${error.message}`,
      });
    }
    throw error;
  }
}

export async function requireAuth(
  request: FastifyRequest,
  reply: FastifyReply
): Promise<string> {
  const uid = await verifyFirebaseToken(request, reply);

  if (!uid) {
    reply.code(401).send({
      error: 'Unauthorized',
      message: 'Authentication required',
    });
    throw new Error('Authentication required');
  }

  return uid;
}
